### ISSUE TITLE
Nan and -inf when calling ./dpmcreate

### ISSUE THREAD

**Issue by rodrigob** [liuliu/ccv#71]
I am currently trying to use dpmcreate to train a single component, no parts, HOG detector baseline, using INRIA training data.

Training seems to go wrong since it  reports 

```
 - collecting negative examples -- (100%)             
 - tune model 1 constant for -inf
 - collecting negative examples -- (100%) 
 - tune model 1 constant for nan
```

Which I suspect to be very wrong.

Full outputs as follows

```
 ./dpmcreate --positive-list ~/datasets/inria_positives_annotations_fullpath_x_y_width_height.txt --background-list ~/datasets/inria_negatives.txt --negative-count 5000 --model-component 1 --model-part 0 --working-dir ./dpm_cache --symmetric 0
with 1237 positive examples and 5000 negative examples
negative examples are are going to be collected from 1218 background images
use symmetric property? no
use color? yes
negative examples cache size : 2000
1 components and 0 parts
expected 20 root relabels, 10 relabels, 50 data minings and 1000 iterations
include overlap : 0.700000
alpha : 0.010000
alpha decreasing ratio : 0.995000
C : 0.002000
balance ratio : 1.500000
------------------------
global mean: 0.339597, & variance: 0.000022
interclass mean(variance): 0.339597(0.000022)
computing root mixture model dimensions: 5x15
root mixture model initialization corrupted, reboot
 - generating positive examples for model 0 : 1237 / 1237
 - generating negative examples for all models : 5000 / 5000
initializing root mixture model for model 1(1)
 - creating initial model 1 at 5x15
 - converting examples to liblinear format: 6236 / 6237     
 - generated 6236 examples with 2326 dimensions each
 - running liblinear for initial linear SVM model (L2-regularized, L1-loss)
.*
optimization finished, #iter = 11
Objective value = -2.434024
nSV = 1686
 - model->label[0]: 1, model->nr_class: 2, model->nr_feature: 2325
components == 1, skipped coordinate-descent to optimize root mixture model
initializing part filters
 - initializing part filters for model 1(1)
optimizing root filter & part filters with stochastic gradient descent
 - collecting responses from positive examples : 100%  
 - positive examples divided by components : 1192
 - collecting negative examples -- (100%)             
 - tune model 1 constant for -inf
 - collecting negative examples -- (100%) 
 - tune model 1 constant for nan
```

I stopped the training since -inf and nan indicate that something is very wrong.

Using  model-part bigger than 0 generates a segfault.

I am using the stable release of libccv  liuliu-ccv-70f8126.

Any idea of what is going on ?

**Comment by liuliu** [comment]
Hey, I am not sure if dpmcreate works with single part model, or 0 part model. But at my point, I did introduce a regression in handling liblinear output https://github.com/liuliu/ccv/commit/98f8d3e6a8dc2c48ce041f479287759d0970144d which makes ccv 0.5 cannot give correct dpmcreate output. This is fixed in stable branch and r0.5-rc1 now.

**Comment by rodrigob** [comment]
Thanks for the fix, the patched version learned a model properly (although its quality was underwhelming)

**Comment by liuliu** [comment]
@rodrigob, is it with parts or no parts? The HOG paper uses padded input from INRIA, whereas the parts training I do uses non-padding version, that may explain difference, otherwise it should get the same result as the HOG paper for no-parts.

**Comment by rodrigob** [comment]
Zero parts.  I did not run  a proper evaluation, but just by visual inspection it was clear that quality is bad. Also training was ridiculously slow. There is code out there (guess which one :P ) that can provide significantly better quality with less than 20 minutes of training... (we will be doing an update release in the next weeks).

**Comment by liuliu** [comment]
@rodrigo looking forward for your new ICF-based training (I think, in retrospective my implementation suffers more overfitting is due to having more features whereas yours uses only square one and sampled at 4 pixel aside).

This dpmcreate uses stochastic gradient descent to optimize the parts model, if zero parts and still goes through that phase, it will be ridiculously slow. Otherwise, it just uses liblinear to train the HOG whole model, which should be relatively fast. The original HOG model is very bad comparing to yours (about 20% less accurate), also, if you don't pad the INRIA rectangle input (which this dpm implementation doesn't), you will get another 10% penalty for that (as the HOG paper suggests). I guess that's why the result to you is pretty bad.

My trained parts model for DPM should be only about 5%~9% short from your ICF implementation though.

**Comment by rodrigob** [comment]
Clarification: (as indicated) in all of our papers we use rectangles as features for the integral channels classifiers. Only in CVPR2013 we mentioned the so-called "SquaresChnftrs" which is a baseline that uses only squares. We explicitly make the point that using all rectangles is (slightly) better than using just squares (figure 3a). Our final result (so called "Roerei" detector) always use rectangles (not only squares).

SquaresChnftrs are recommended as a good trade-off between speed and quality.

**Comment by liuliu** [comment]
Thanks for the clarification! I skimmed the paper and thought for the exhaust feature searching part you used square feature. Interested to see your code to know why my ICF training suffered from significant overfitting on larger models (in terms of dimensions).
